File: EvalVisitor.py
import math
return 0


# Cambiar modo de ángulos
def visitSetMode(self, ctx):
mode = ctx.getChild(2).getText()
self.useDegrees = (mode == "deg")
print("Modo cambiado a:", mode)
return 0


# Variables
def visitId(self, ctx):
id = ctx.ID().getText()
return self.memory.get(id, 0)


# Números
def visitNumber(self, ctx):
return float(ctx.NUMBER().getText())


# Paréntesis
def visitParens(self, ctx):
return self.visit(ctx.expr())


# Suma y resta
def visitAddSub(self, ctx):
left = self.visit(ctx.expr(0))
right = self.visit(ctx.expr(1))
if ctx.op.type == LabeledExprParser.PLUS:
return left + right
return left - right


# Multiplicación y división
def visitMulDiv(self, ctx):
left = self.visit(ctx.expr(0))
right = self.visit(ctx.expr(1))
if ctx.op.text == '*':
return left * right
return left / right


# Potencia
def visitPower(self, ctx):
return math.pow(self.visit(ctx.expr(0)), self.visit(ctx.expr(1)))


# Factorial
def visitFactorial(self, ctx):
return factorial(int(self.visit(ctx.expr())))


# Funciones matemáticas
def visitFunction(self, ctx):
func = ctx.func.text
arg = self.visit(ctx.expr())


if self.useDegrees and func in ("Sin", "Cos", "Tan"):
arg = math.radians(arg)


if func == "Sin": return math.sin(arg)
if func == "Cos": return math.cos(arg)
if func == "Tan": return math.tan(arg)
if func == "Sqrt": return math.sqrt(arg)
if func == "Ln": return math.log(arg)
if func == "Log": return math.log10(arg)
if func == "Rad": return math.radians(arg)
if func == "Deg": return math.degrees(arg)
return 0

File: EvalVisitor.java
import java.util.*;
return visit(ctx.expr());
}


// Suma y resta
@Override
public Double visitAddSub(LabeledExprParser.AddSubContext ctx) {
double left = visit(ctx.expr(0));
double right = visit(ctx.expr(1));
return ctx.op.getType() == LabeledExprParser.PLUS ? left + right : left - right;
}


// Multiplicación y división
@Override
public Double visitMulDiv(LabeledExprParser.MulDivContext ctx) {
double left = visit(ctx.expr(0));
double right = visit(ctx.expr(1));
return ctx.op.getText().equals("*") ? left * right : left / right;
}


// Potencia
@Override
public Double visitPower(LabeledExprParser.PowerContext ctx) {
double base = visit(ctx.expr(0));
double exp = visit(ctx.expr(1));
return Math.pow(base, exp);
}


// Factorial
@Override
public Double visitFactorial(LabeledExprParser.FactorialContext ctx) {
double val = visit(ctx.expr());
return factorial((int) val);
}


private double factorial(int n) {
if (n <= 1) return 1;
return n * factorial(n - 1);
}


// Funciones matemáticas (sin, cos, sqrt, etc.)
@Override
public Double visitFunction(LabeledExprParser.FunctionContext ctx) {
String func = ctx.func.getText();
double arg = visit(ctx.expr());


// Convertir a radianes si estamos en modo grados
if (useDegrees && (func.equals("Sin") || func.equals("Cos") || func.equals("Tan"))) {
arg = Math.toRadians(arg);
}


switch (func) {
case "Sin": return Math.sin(arg);
case "Cos": return Math.cos(arg);
case "Tan": return Math.tan(arg);
case "Sqrt": return Math.sqrt(arg);
case "Ln": return Math.log(arg);
case "Log": return Math.log10(arg);
case "Rad": return Math.toRadians(arg);
case "Deg": return Math.toDegrees(arg);
default: return 0.0;
}
}
}
